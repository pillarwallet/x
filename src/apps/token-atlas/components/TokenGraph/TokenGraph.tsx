import {
  CategoryScale,
  ChartArea,
  ChartData,
  Chart as ChartJS,
  ChartOptions,
  CoreScaleOptions,
  Filler,
  Legend,
  LineElement,
  LinearScale,
  PointElement,
  Scale,
  ScriptableContext,
  TimeScale,
  Tooltip,
  TooltipItem,
} from 'chart.js';
import 'chartjs-adapter-date-fns';
import { useCallback } from 'react';
import { Line } from 'react-chartjs-2';

// hooks
import { useAppSelector } from '../../hooks/useReducerHooks';

// types
import {
  MarketHistoryPairData,
  TokenAtlasInfoData,
} from '../../../../types/api';
import { PeriodFilter } from '../../types/types';

// components
import SkeletonLoader from '../../../../components/SkeletonLoader';
import Body from '../Typography/Body';

type Scales = {
  [key: string]: Scale<CoreScaleOptions>;
};

ChartJS.register(
  LineElement,
  PointElement,
  CategoryScale,
  LinearScale,
  TimeScale,
  Filler,
  Tooltip,
  Legend
);

const TokenGraph = () => {
  const tokenDataGraph = useAppSelector(
    (state) =>
      state.tokenAtlas.tokenDataGraph as MarketHistoryPairData | undefined
  );
  const tokenDataInfo = useAppSelector(
    (state) => state.tokenAtlas.tokenDataInfo as TokenAtlasInfoData | undefined
  );
  const periodFilter = useAppSelector(
    (state) => state.tokenAtlas.periodFilter as PeriodFilter
  );
  const isGraphLoading = useAppSelector(
    (state) => state.tokenAtlas.isGraphLoading as boolean
  );
  const isGraphErroring = useAppSelector(
    (state) => state.tokenAtlas.isGraphErroring as boolean
  );

  // This gets the right color depending on the price value threshold, a callback is used
  // to make sure this only gets called if tokenDataInfo.price changes, otherwise the graph crashes
  const getGradient = useCallback(
    (ctx: CanvasRenderingContext2D, chartArea: ChartArea, scales: Scales) => {
      const gradientBg = ctx.createLinearGradient(
        0,
        chartArea.top,
        0,
        chartArea.bottom
      );
      const percentage = Math.min(
        Math.max(
          (scales.y.getPixelForValue(tokenDataInfo?.price || 0) -
            chartArea.top) /
            chartArea.height,
          0
        ),
        1
      );

      // this changes the color of the line depending on a value threshold
      gradientBg.addColorStop(0, '#6CFF00');
      gradientBg.addColorStop(percentage, '#6CFF00');
      gradientBg.addColorStop(percentage, '#FF005C');
      gradientBg.addColorStop(1, '#FF005C');

      return gradientBg;
    },
    [tokenDataInfo?.price]
  );

  // This gets the right graph time scale (x) depending on the periodFilter selected. A callback is used
  // to make sure this only gets called if tokenDataGraph?.result.data changes, otherwise the graph crashes
  const graphXScale = useCallback(
    () => {
      if (tokenDataGraph?.result.data.length) {
        if (periodFilter === PeriodFilter.HOUR) {
          return 'minute';
        }

        if (periodFilter === PeriodFilter.DAY) {
          return 'hour';
        }

        if (periodFilter === PeriodFilter.WEEK) {
          return 'day';
        }

        if (periodFilter === PeriodFilter.MONTH) {
          return 'day';
        }

        if (periodFilter === PeriodFilter.YEAR) {
          return 'month';
        }

        return 'day';
      }

      return 'day';
    },
    [tokenDataGraph?.result.data.length] // eslint-disable-line react-hooks/exhaustive-deps
  );

  if (isGraphLoading) {
    return <SkeletonLoader $height="300px" $radius="6px" />;
  }

  if (!tokenDataGraph?.result.data.length && !isGraphErroring) {
    return <Body className="my-4">Price history not found.</Body>;
  }

  if (!tokenDataGraph?.result.data.length && isGraphErroring) {
    return (
      <Body className="my-4">
        Oops something went wrong! The data for this timeframe may not be
        available. Please try selecting a different timeframe or reload the
        page.
      </Body>
    );
  }

  const createGradient = (
    context: ScriptableContext<'line'>
  ): CanvasGradient | undefined => {
    const { chart } = context;
    const { ctx, chartArea, scales } = chart;

    if (!chartArea) return undefined;

    return getGradient(ctx, chartArea, scales);
  };

  // This gives us the right dataset for the graph
  const data: ChartData<'line'> = {
    labels: tokenDataGraph?.result.data.map((x) => x.time),
    datasets: [
      {
        label: `${
          tokenDataInfo?.symbol ? tokenDataInfo?.symbol : 'Token'
        } price`,
        data: tokenDataGraph?.result.data.map((price) => price.close) || [],
        borderColor: (ctx: ScriptableContext<'line'>) => createGradient(ctx),
        tension: 0.4,
        pointRadius: 0,
      },
    ],
  };

  // This draws the price value threshold dotted line
  const horizontalDottedLine = {
    id: 'horizontalDottedLine',
    beforeDatasetDraw(chart: ChartJS) {
      const {
        ctx,
        chartArea: { left, width },
        scales: { y },
      } = chart;
      ctx.save();

      ctx.strokeStyle = 'white';
      ctx.setLineDash([3, 6]);
      ctx.lineWidth = 2;
      ctx.strokeRect(
        left,
        y.getPixelForValue(tokenDataInfo?.price || 0),
        width,
        0
      );
      ctx.restore();
    },
  };

  // The options used to customize the UI of the graph
  const options: ChartOptions<'line'> = {
    responsive: true,
    maintainAspectRatio: false,
    plugins: {
      title: {
        display: true,
        text: 'Token price history',
      },
      filler: {
        drawTime: 'beforeDatasetDraw',
        propagate: true,
      },
      legend: {
        display: false,
      },
      tooltip: {
        callbacks: {
          label: (tooltipItem: TooltipItem<'line'>) => {
            let label = tooltipItem.dataset.label || '';
            if (label) {
              label += ': ';
            }
            label += parseFloat(tooltipItem.parsed.y.toString()).toFixed(8);
            return label;
          },
        },
      },
    },
    scales: {
      x: {
        offset: false,
        type: 'time',
        time: {
          unit: graphXScale() || 'day',
          displayFormats: {
            minute: 'HH:mm',
            hour: 'HH:mm',
            day: 'dd MMM',
            month: 'MMM y',
          },
        },
        border: {
          width: 0,
        },
        grid: {
          drawOnChartArea: false,
          drawTicks: false,
        },
        ticks: {
          font: {
            weight: 600,
            size: 9,
          },
          stepSize:
            periodFilter === PeriodFilter.HOUR ||
            periodFilter === PeriodFilter.DAY ||
            periodFilter === PeriodFilter.MONTH
              ? 3
              : 1,
          color: '#DDDDDD',
        },
      },
      y: {
        border: {
          width: 0,
        },
        grid: {
          drawOnChartArea: false,
          drawTicks: false,
        },
        position: 'right',
        ticks: {
          showLabelBackdrop: true,
          backdropColor: (ctx) =>
            ctx.tick.value === (tokenDataInfo?.price || 0)
              ? '#6CFF00'
              : 'rgba(49, 49, 49, 0.5)',
          backdropPadding: 3,
          font: {
            weight: 600,
            size: 9,
          },
          color: (ctx) =>
            ctx.tick.value === (tokenDataInfo?.price || 0)
              ? '#313131'
              : '#DDDDDD',
        },
      },
    },
    interaction: {
      mode: 'index',
      intersect: false,
    },
  };

  return (
    <div
      id="token-atlas-token-graph"
      className="flex w-[99%] mb-20 h-full max-h-[400px] mobile:mb-0"
    >
      <Line
        data={data}
        options={options}
        plugins={[horizontalDottedLine]}
        data-testid="price-graph"
      />
    </div>
  );
};

export default TokenGraph;
