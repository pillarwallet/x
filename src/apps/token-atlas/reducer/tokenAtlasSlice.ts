/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { sub } from 'date-fns';

// types
import { Token } from '../../../services/tokensData';
import {
  BlockchainData,
  TokenAtlasInfoData,
  TokenMarketHistory,
  TokenPriceGraphPeriod,
} from '../../../types/api';
import { ChainType, PeriodFilter, SelectedTokenType } from '../types/types';

// utils
import { convertDateToUnixTimestamp } from '../../../utils/common';

export type TokenAltasState = {
  isSearchTokenModalOpen: boolean;
  isSelectChainDropdownOpen: boolean;
  tokenListData: Token[];
  selectedChain: ChainType;
  searchTokenResult: Token[];
  selectedToken: SelectedTokenType | undefined;
  tokenDataInfo: TokenAtlasInfoData | undefined;
  tokenDataGraph: TokenMarketHistory | undefined;
  isAllChainsVisible: boolean;
  priceGraphPeriod: TokenPriceGraphPeriod;
  periodFilter: PeriodFilter;
  isGraphLoading: boolean;
  blockchainList: BlockchainData[] | undefined;
  searchToken: string;
  isTokenSearchLoading: boolean;
  isTokenSearchErroring: boolean;
};

const initialState: TokenAltasState = {
  isSearchTokenModalOpen: false,
  isSelectChainDropdownOpen: false,
  tokenListData: [],
  selectedChain: { chainId: 0, chainName: 'all' },
  searchTokenResult: [],
  selectedToken: undefined,
  tokenDataInfo: undefined,
  tokenDataGraph: undefined,
  isAllChainsVisible: false,
  priceGraphPeriod: {
    from: convertDateToUnixTimestamp(sub(new Date(), { days: 1 })),
    to: undefined,
  },
  periodFilter: PeriodFilter.DAY,
  isGraphLoading: false,
  blockchainList: [],
  searchToken: '',
  isTokenSearchLoading: false,
  isTokenSearchErroring: false,
};

const tokenAtlasSlice = createSlice({
  name: 'tokenAtlas',
  initialState,
  reducers: {
    setIsSearchTokenModalOpen(state, action: PayloadAction<boolean>) {
      state.isSearchTokenModalOpen = action.payload;
    },
    setIsSelectChainDropdownOpen(state, action: PayloadAction<boolean>) {
      state.isSelectChainDropdownOpen = action.payload;
    },
    setTokenListData(state, action: PayloadAction<Token[]>) {
      state.tokenListData = action.payload;
    },
    setSelectedChain(state, action: PayloadAction<ChainType>) {
      state.selectedChain = action.payload;
    },
    setSearchTokenResult(state, action: PayloadAction<Token[]>) {
      state.searchTokenResult = action.payload;
    },
    setSelectedToken(
      state,
      action: PayloadAction<SelectedTokenType | undefined>
    ) {
      state.selectedToken = action.payload;
    },
    setTokenDataInfo(
      state,
      action: PayloadAction<TokenAtlasInfoData | undefined>
    ) {
      state.tokenDataInfo = action.payload;
    },
    setTokenDataGraph(
      state,
      action: PayloadAction<TokenMarketHistory | undefined>
    ) {
      state.tokenDataGraph = action.payload;
    },
    setIsAllChainsVisible(state, action: PayloadAction<boolean>) {
      state.isAllChainsVisible = action.payload;
    },
    setPriceGraphPeriod(state, action: PayloadAction<TokenPriceGraphPeriod>) {
      state.priceGraphPeriod = action.payload;
    },
    setPeriodFilter(state, action: PayloadAction<PeriodFilter>) {
      state.periodFilter = action.payload;
    },
    setIsGraphLoading(state, action: PayloadAction<boolean>) {
      state.isGraphLoading = action.payload;
    },
    setBlockchainList(
      state,
      action: PayloadAction<BlockchainData[] | undefined>
    ) {
      state.blockchainList = action.payload;
    },
    setSearchToken(state, action: PayloadAction<string>) {
      state.searchToken = action.payload;
    },
    setIsTokenSearchLoading(state, action: PayloadAction<boolean>) {
      state.isTokenSearchLoading = action.payload;
    },
    setIsTokenSearchErroring(state, action: PayloadAction<boolean>) {
      state.isTokenSearchErroring = action.payload;
    },
  },
});

export const {
  setIsSearchTokenModalOpen,
  setIsSelectChainDropdownOpen,
  setTokenListData,
  setSelectedChain,
  setSearchTokenResult,
  setSelectedToken,
  setTokenDataInfo,
  setTokenDataGraph,
  setIsAllChainsVisible,
  setPriceGraphPeriod,
  setPeriodFilter,
  setIsGraphLoading,
  setBlockchainList,
  setSearchToken,
  setIsTokenSearchLoading,
  setIsTokenSearchErroring,
} = tokenAtlasSlice.actions;

export default tokenAtlasSlice;
