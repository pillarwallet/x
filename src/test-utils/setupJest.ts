/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { TextDecoder, TextEncoder } from 'util';

// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { BigNumber } from 'ethers';
import 'jest-styled-components';
import { createWalletClient, http } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { goerli } from 'viem/chains';

// mocking matchMedia for slick carousel
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// polyfill for TextEncoder and TextDecoder for jsdom environment (viem dep related)
Object.assign(global, { TextDecoder, TextEncoder });

jest.mock('@firebase/app');
jest.mock('@firebase/analytics');
jest.mock('axios');
jest.mock('@etherspot/data-utils');
jest.mock('@etherspot/modular-sdk');

jest.mock('@privy-io/react-auth', () => ({
  PrivyProvider: ({ children }: { children: React.ReactNode }) => children,
  usePrivy: jest.fn(() => ({ authenticated: false })),
  useWallets: jest.fn(() => ({})),
  useLogout: jest.fn(() => ({ logout: jest.fn() })),
}));

export const etherspotTestAssets = [
  {
    address: '0xe3818504c1b32bf1557b16c238b2e01fd3149c17',
    chainId: 1,
    name: 'tk1',
    symbol: 'TK1',
    decimals: 18,
    logoURI: '',
  },
  {
    address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
    chainId: 1,
    name: 'tk2',
    symbol: 'TK2',
    decimals: 18,
    logoURI: '',
  },
  {
    address: '0xdac17f958d2ee523a2206206994597c13d831ec7',
    chainId: 1,
    name: 'tk3',
    symbol: 'TK3',
    decimals: 18,
    logoURI: '',
  },
];

export const etherspotTestSupportedAssets = [
  {
    address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
    chainId: 1,
    decimals: 6,
    icon: '',
    name: 'USDC',
    symbol: 'USDC',
  },
  {
    address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
    chainId: 1,
    decimals: 18,
    icon: '',
    name: 'Wrapped ETH',
    symbol: 'WETH',
  },
  {
    address: '0x39B061B7e41DE8B721f9aEcEB6b3f17ECB7ba63E',
    chainId: 420,
    decimals: 18,
    icon: '',
    name: 'nextWETH',
    symbol: 'nextWETH',
  },
];

const randomWallet = privateKeyToAccount(
  `0x${crypto.getRandomValues(new Uint8Array(32)).reduce((acc, byte) => acc + byte.toString(16).padStart(2, '0'), '')}`
);
const provider = createWalletClient({
  account: randomWallet,
  chain: goerli,
  transport: http('http://localhost:8545'),
});

jest.mock('wagmi', () => ({
  createConfig: jest.fn(),
  http: jest.fn(),
  mainnet: { id: 1, name: 'mainnet' },
  useAccount: jest.fn().mockReturnValue({
    address: '0x',
  }),
  useDisconnect: jest.fn().mockReturnValue({
    disconnect: jest.fn(),
  }),
}));

jest.mock('wagmi/connectors', () => ({
  walletConnect: jest.fn(),
}));

jest.mock('@etherspot/transaction-kit', () => ({
  useEtherspotAssets: jest.fn().mockReturnValue({
    getAssets: async () => etherspotTestAssets,
    getSupportedAssets: async () => etherspotTestSupportedAssets,
  }),
  useWalletAddress: jest
    .fn()
    .mockReturnValue('0x7F30B1960D5556929B03a0339814fE903c55a347'),
  useEtherspotTransactions: jest.fn().mockReturnValue({
    chainId: 1,
    batches: [],
    estimate: async () => [],
    send: async () => [
      {
        sentBatches: [{ userOpHash: '0x123' }],
        estimatedBatches: [],
        batches: [],
      },
    ],
    isEstimating: false,
    isSending: false,
    containsEstimatingError: false,
    containsSendingError: false,
  }),
  useEtherspotBalances: jest.fn().mockReturnValue({
    getAccountBalances: async () => [],
  }),
  useEtherspotPrices: jest.fn().mockReturnValue({
    getPrice: async () => undefined,
    getPrices: async () => [],
  }),

  useEtherspotHistory: jest.fn().mockReturnValue({
    getAccountTransactions: async () => [],
    getAccountTransaction: async () => undefined,
    getAccountTransactionStatus: async () => undefined,
  }),

  useEtherspotNfts: jest.fn().mockReturnValue({
    getAccountNfts: async () => [],
  }),

  useEtherspot: jest.fn().mockReturnValue({
    getSdk: async () => {},
    getDataService: () => {},
    provider,
    chainId: 1,
  }),

  useEtherspotUtils: jest.fn().mockReturnValue({
    checksumAddress: () => '0x7F30B1960D5556929B03a0339814fE903c55a347',
    verifyEip1271Message: async () => false,
    toBigNumber: () => BigNumber.from('1'),
    parseBigNumber: () => '0x123',
    isZeroAddress: () => false,
    addressesEqual: () => true,
  }),
}));

process.env.REACT_APP_PRIVY_APP_ID = 'test';
