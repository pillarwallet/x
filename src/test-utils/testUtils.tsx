import { configureStore } from '@reduxjs/toolkit';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

// store
import { store } from '../store';

// theme
import { defaultTheme } from '../theme';

// reducers
import depositSlice from '../apps/deposit/reducer/depositSlice';
import leaderboardSlice from '../apps/leaderboard/reducer/LeaderboardSlice';
import walletPortfolioSlice from '../apps/pillarx-app/reducer/WalletPortfolioSlice';
import swapSlice from '../apps/the-exchange/reducer/theExchangeSlice';
import tokenAtlasSlice from '../apps/token-atlas/reducer/tokenAtlasSlice';

// types
import { PeriodFilter } from '../apps/token-atlas/types/types';

// providers
import AccountTransactionHistoryProvider from '../providers/AccountTransactionHistoryProvider';
import BottomMenuModalProvider from '../providers/BottomMenuModalProvider';
import GlobalTransactionsBatchProvider from '../providers/GlobalTransactionsBatchProvider';
import LanguageProvider from '../providers/LanguageProvider';
import { PrivateKeyLoginProvider } from '../providers/PrivateKeyLoginProvider';
import SelectedChainsHistoryProvider from '../providers/SelectedChainsHistoryProvider';

// Create a test store with all reducers
const testStore = configureStore({
  reducer: {
    deposit: depositSlice.reducer,
    leaderboard: leaderboardSlice.reducer,
    walletPortfolio: walletPortfolioSlice.reducer,
    swap: swapSlice.reducer,
    tokenAtlas: tokenAtlasSlice.reducer,
  },
  preloadedState: {
    deposit: depositSlice.getInitialState(),
    leaderboard: leaderboardSlice.getInitialState(),
    walletPortfolio: walletPortfolioSlice.getInitialState(),
    swap: swapSlice.getInitialState(),
    tokenAtlas: {
      ...tokenAtlasSlice.getInitialState(),
      tokenDataGraph: {
        result: {
          data: [
            {
              volume: 1050.25,
              open: 45000.5,
              high: 45500.0,
              low: 44800.25,
              close: 45250.75,
              time: 1712457600,
            },
            {
              volume: 980.1,
              open: 45250.75,
              high: 46000.0,
              low: 45000.0,
              close: 45900.0,
              time: 1712544000,
            },
            {
              volume: 1125.8,
              open: 45900.0,
              high: 46250.5,
              low: 45700.0,
              close: 46050.5,
              time: 1712630400,
            },
          ],
        },
      },
      tokenDataInfo: {
        id: 1,
        market_cap: 100,
        market_cap_diluted: 100,
        liquidity: 150,
        price: 105,
        off_chain_volume: 150,
        volume: 150,
        volume_change_24h: 1.54,
        volume_7d: 2.3,
        is_listed: true,
        price_change_24h: 0.44,
        price_change_1h: 0.1,
        price_change_7d: 1.8,
        price_change_1m: 3.4,
        price_change_1y: 6.7,
        ath: 146,
        atl: 96,
        name: 'TOKEN',
        symbol: 'TKN',
        logo: 'tokenLogo.png',
        rank: 1047,
        contracts: [],
        total_supply: '300',
        circulating_supply: '170',
      },
      periodFilter: PeriodFilter.DAY,
      isGraphLoading: false,
      isGraphErroring: false,
    },
  },
});

// Test wrapper with all necessary providers
export const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <Provider store={testStore}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>
          <AccountTransactionHistoryProvider>
            <BottomMenuModalProvider>
              <GlobalTransactionsBatchProvider>
                <SelectedChainsHistoryProvider>
                  <PrivateKeyLoginProvider>{children}</PrivateKeyLoginProvider>
                </SelectedChainsHistoryProvider>
              </GlobalTransactionsBatchProvider>
            </BottomMenuModalProvider>
          </AccountTransactionHistoryProvider>
        </LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);

// Test wrapper with only Redux and Router (for simpler components)
export const SimpleTestWrapper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => (
  <Provider store={store}>
    <BrowserRouter>{children}</BrowserRouter>
  </Provider>
);

// Test wrapper with Redux, Router, and basic providers
export const BasicTestWrapper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => (
  <Provider store={store}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>{children}</LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);

// Test wrapper for exchange components
export const ExchangeTestWrapper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => (
  <Provider store={store}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>
          <GlobalTransactionsBatchProvider>
            {children}
          </GlobalTransactionsBatchProvider>
        </LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);

// Test wrapper for token atlas components
export const TokenAtlasTestWrapper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => (
  <Provider store={store}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>
          <GlobalTransactionsBatchProvider>
            {children}
          </GlobalTransactionsBatchProvider>
        </LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);

// Test wrapper for leaderboard components
export const LeaderboardTestWrapper: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => (
  <Provider store={store}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>
          <GlobalTransactionsBatchProvider>
            {children}
          </GlobalTransactionsBatchProvider>
        </LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);

// Test wrapper for deposit components
export const DepositTestWrapper: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => (
  <Provider store={store}>
    <BrowserRouter>
      <ThemeProvider theme={defaultTheme}>
        <LanguageProvider>
          <GlobalTransactionsBatchProvider>
            {children}
          </GlobalTransactionsBatchProvider>
        </LanguageProvider>
      </ThemeProvider>
    </BrowserRouter>
  </Provider>
);
