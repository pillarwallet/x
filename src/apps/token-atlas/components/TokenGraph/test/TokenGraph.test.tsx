import { configureStore } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

// test utils
import { TestWrapper } from '../../../../../test-utils/testUtils';

// types

// components
import TokenGraph from '../TokenGraph';

// reducer
import tokenAtlasSlice from '../../../reducer/tokenAtlasSlice';

describe('<TokenGraph />', () => {
  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <TokenGraph />
        </TestWrapper>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders the graph with data when available', () => {
    render(
      <TestWrapper>
        <TokenGraph />
      </TestWrapper>
    );

    expect(screen.getByTestId('price-graph')).toBeInTheDocument();
  });

  it('renders the graph container with correct styling', () => {
    render(
      <TestWrapper>
        <TokenGraph />
      </TestWrapper>
    );

    const container = screen.getByTestId('price-graph').parentElement;
    expect(container).toHaveAttribute('id', 'token-atlas-token-graph');
    expect(container).toHaveClass(
      'flex',
      'w-[99%]',
      'mb-20',
      'h-full',
      'max-h-[400px]',
      'mobile:mb-0'
    );
  });

  it('displays a message when no price history is available', () => {
    // Create a custom test store with empty data
    const testStore = configureStore({
      reducer: {
        tokenAtlas: tokenAtlasSlice.reducer,
      },
      preloadedState: {
        tokenAtlas: {
          ...tokenAtlasSlice.getInitialState(),
          tokenDataGraph: { result: { data: [] } },
          isGraphLoading: false,
          isGraphErroring: false,
        },
      },
    });

    render(
      <Provider store={testStore}>
        <TokenGraph />
      </Provider>
    );

    expect(screen.getByText('Price history not found.')).toBeInTheDocument();
  });
});
