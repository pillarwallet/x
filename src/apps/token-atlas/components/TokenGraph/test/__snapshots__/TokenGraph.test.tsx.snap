// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TokenGraph /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex w-[99%] mb-20 h-full max-h-[400px] mobile:mb-0"
  id="token-atlas-token-graph"
>
  <canvas
    data-testid="price-graph"
    height="150"
    plugins={
      [
        {
          "beforeDatasetDraw": [Function],
          "id": "horizontalDottedLine",
        },
      ]
    }
    role="img"
    width="300"
  />
</div>
`;
